lib/chef/knife/cleanup_versions.rb

```diff
@@ -94,13 +98,53 @@ def cookbooks
         end
         print "\n"
       end
-      
+
       if !config[:delete]
         ui.msg "Not deleting unused cookbook versions; use --delete if you want to remove them"
       end
-      
     end
-    
+
+    def keep_for_runlist(cb_versions, env_name, runlist)
+      runlist_req = { "run_list"  => [ runlist ] }
+      begin
+        run_cookbooks = \
+          rest.post_rest("/environments/#{env_name}/cookbook_versions", runlist_req)
```

Comment:
Some runlists simply can't be evaluated in some environments if the constraints can't be satisfied.

I've not merged this in because I'd like to get some test conditions around it. 

Try this to test:

```javascript
cat<<END> runlist.json
{
  "run_list": [
    "recipe[apt]"
  ]
}
END
knife raw -i runlist.json -m POST environments/my_env/cookbook_versions
```

Also, I'm pretty sure my examples with roles doesn't work now that I've had another look.  That said, it does work with cookbooks as the runlist, and, further, it doesn't do __any__ modifications untill you pass the --delete option